dataclasses_json/core.py

```diff
@@ -158,17 +158,18 @@ def _decode_dataclass(cls, kvs, infer_missing):
 
         field_value = kvs[field.name]
         field_type = types[field.name]
-        if field_value is None and not _is_optional(field_type):
-            warning = (f"value of non-optional type {field.name} detected "
-                       f"when decoding {cls.__name__}")
-            if infer_missing:
-                warnings.warn(
-                    f"Missing {warning} and was defaulted to None by "
-                    f"infer_missing=True. "
-                    f"Set infer_missing=False (the default) to prevent this "
-                    f"behavior.", RuntimeWarning)
-            else:
-                warnings.warn(f"`NoneType` object {warning}.", RuntimeWarning)
+        if field_value is None:
+            if not _is_optional(field_type):
```

Comment:
Any particular reason to split the if-statement and introduce further nesting?

Reply:
The line at 172 (old code) `init_kwargs[field.name] = field_value` is only hit if the value is None *and* `field_type` is not optional. If anything, it should update `init_kwargs` when `field_type` is optional.

Reply:
Apologies, I missed the part that `init_kwargs[field.name] = field_value` was not indented as well.